working/0546-patterns/goals-and-constraints.md

```diff
@@ -0,0 +1,461 @@
+}
+// Prints "two".
+```
+
+This introduces significant complexity. We'll also want variable binding
+patterns, which are also naturally represented as an identifier. If we allow
+constant patterns to also be simple bare identifiers, it means we need to do
+lexical scope resolution to determine if a pattern is a constant pattern or a
+variable pattern.
+
+We'll presumably want to support dotted identifiers for enum cases. I imagine
+users will expect that to also work for named constants imported from libraries
+with a prefix.
+
+That raises the question of whether we want to support arbitrary constant
+expressions. That has some appeal, but means that much of the expression grammar
+would overlap the pattern grammar, which leads to ambiguity and pain. Consider,
+for example, `[1, 2]`. Is that a constant pattern containing a constant list
+literal, or a list destructuring pattern containing two literal patterns?
+
+### Destructure collections
+
+A key goal and a feature offered by other languages is patterns that can match
+and pull elements out of collections. Dart already has lists, maps, and sets.
+Patterns to let you destructure those are an obvious feature.
+
+We also want to add [tuples][] and those almost require destructuring because it's
+difficult to otherwise access the elements in a type safe way, since tuples are
+heterogeneously typed.
+
+[tuples]: https://github.com/dart-lang/language/issues/68
+
+As in other languages, the syntax for these should mirror the corresponding
+collection literal expression form:
+
+```dart
+Object obj = ...
+match (obj) {
+  case [a, b] => print("list with elements $a and $b");
+  case {a, b} => print("set with elements $a and $b");
+  case (a, b) => print("tuple with elements $a and $b");
+}
+```
+
+Maps are a little trickier. Do they match specific entries, in which case the
+key is a value, or do they destructure any entry, in which case the key is a
+pattern?
+
+### Destructure instances of classes
+
+Another common use of pattern matching is for working with [sum types][]. In
+typed functional languages, sum types are the typical way to define values that
+can be one of a few different cases, with potentially nested data for each case.
+Pattern matching is then how you implement behavior that varies based on which
+case you have.
+
+[sum types]: https://en.wikipedia.org/wiki/Tagged_union
+
+In an object-oriented language, those cases are modeled using subclasses. Thus,
+to use pattern matching to select behavior based on a sum type case, we need
+patterns that match a specific subclass and extract fields from it.
+
+For example, something like this:
+
+```dart
+abstract class Shape {}
+class Rect extends Shape {
+  final num left, top, right, bottom;
+  Rect(this.left, this.top, this.right, this.bottom);
+}
+class Circle extends Shape {
+  final num x, y, radius;
+  Circle(this.x, this.y, this.radius);
+}
+
+printShape(Shape shape) {
+  match (shape) {
+    case Rect(left, top, right, bottom) =>
```

Comment:
What C# does here is: `Rect{Left: var left, Top: var top, Right: var right, Bottom: var bottom}`.
This separates the property name from the variable name, which is necessary when you can also pattern match the individual property values:
```c#
if (shape is Rect(Left: 0, Bottom: 0, Right: var width, Top: var height) rect) {
  ...
}
```
It's necessary to distinguish `Left: x` from `Left: var x` because the former uses the existing variable `x` as test value and the latter introduces a new `x` variable with a non-refutable pattern ([Example](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQKgLwPYIMpwAOhuMATADQAmIA1AD4ACZAjALABQnjAzAATMBLAOycA3pz5SBZMpOkSO05XwCSeABZRCEPgj4BePgDsIAdz4AlCAGMYACgAMFPs74sXZAJQBueSr4ASzA+e31A5CtbGDEAGwgwGBBXF2BsGBhsAFtktwRAgHMNJL4ANyh9M0CqGA0XTMJk8v0NCELigF89aK8+RQCAxhYATnsAIgARaqCSsb5abrsAOnyimHm+MYAPOYXW9phffxUO49OOc8DjGAgEMCgbXXUtHT7jq/WAQSQoPoKIGA+c7nQTWOx8ZLPbS6frSXhBa58eKJPxKOH8D56A6o5TwzENHHohHrNIZbKEqTwsEOTHImAuTGkzJZBmI1bFVnrBq9WEBWoRJZ0wxIhKA47KfnIFYHYXssVovkaAVM7LClVZCkqSVLBrCglnY54xHfaBvBUqf7rXkDATCUJyvioEWJXoAKlCuqd6qO5uU5z9nHOQA==)).